### ISSUE TITLE
Invalid Lambda Permission generated for HTTP API $default routes

### ISSUE THREAD

**Issue by txase** [aws/serverless-application-model#1860]
**Description:**

When specifying a `$default` HTTP API route and integration with a Lambda Function, SAM generates an incorrect permission statement. The generated statement has a SourceArn condition that looks like:

```
arn:aws:execute-api:<region>:<accountId>:<apiId>/*/*/$default
```

But this arn has too many `/*`. The correct SourceArn condition looks like:

```
arn:aws:execute-api:<region>:<accountId>:<apiId>/*/$default
```

I have validated this SourceArn condition allows for the integrated Lambda Function to be invoked, and you can also see this source ARN in a screenshot from the HTTP API console describing the required permission:

![image](https://user-images.githubusercontent.com/962436/102108863-48eb0180-3de8-11eb-93df-2fd4a086b1a7.png)

**Steps to reproduce the issue:**
1. Create a CFN stack using this template:
    ```yaml
    Transform: AWS::Serverless-2016-10-31
    Resources:
      Function:
        Type: AWS::Serverless::Function
        Properties:
          InlineCode: 'module.exports.handler = () => ({})'
          Handler: index.handler
          Runtime: nodejs12.x
          Events:
            HttpApiANYdefault:
              Type: HttpApi
              Properties:
                Path: /$default
                Method: ANY
                ApiId: !Ref HttpApi
      HttpApi:
        Type: AWS::Serverless::HttpApi
        Properties:
          DefinitionBody:
            openapi: '3.0'
            info:
              title: !Sub ${AWS::StackName}-HttpApi
              version: '1.0'
            paths:
              /$default:
                x-amazon-apigateway-any-method:
                  responses: {}
                  isDefaultRoute: true
          FailOnWarnings: true
    ```

**Observed result:**

View the generated AWS::Lambda::Permission with the incorrect SourceArn:
```json
    "FunctionHttpApiANYdefaultPermission": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "Principal": "apigateway.amazonaws.com",
        "FunctionName": {
          "Ref": "Function"
        },
        "SourceArn": {
          "Fn::Sub": [
            "arn:${AWS::Partition}:execute-api:${AWS::Region}:${AWS::AccountId}:${__ApiId__}/${__Stage__}/*/$default",
            {
              "__Stage__": "*",
              "__ApiId__": {
                "Ref": "HttpApi"
              }
            }
          ]
        }
      }
    }
```

**Expected result:**

The generated AWS::Lambda::Permission should have the following SourceArn:
```json
    "FunctionHttpApiANYdefaultPermission": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "Principal": "apigateway.amazonaws.com",
        "FunctionName": {
          "Ref": "Function"
        },
        "SourceArn": {
          "Fn::Sub": [
            "arn:${AWS::Partition}:execute-api:${AWS::Region}:${AWS::AccountId}:${__ApiId__}/${__Stage__}/$default",
            {
              "__Stage__": "*",
              "__ApiId__": {
                "Ref": "HttpApi"
              }
            }
          ]
        }
      }
    }
```

**Comment by mndeveci** [comment]
Hi @txase ,

I checked the example template that you provided and as you said the generated ARN has `*` just before the `/$default` path, and it is coming from `Method: ANY` in the `HttpApiANYdefault` event source definition.

Looking at the documentation [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-iam-policy-examples-for-api-execution.html) and [here](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-control-access-using-iam-policies-to-invoke-api.html), it says that `execute-api` ARN can be formed as `"arn:aws:execute-api:region:account-id:api-id/stage/METHOD_HTTP_VERB/Resource-path"`, and the `*` in the example is coming from `ANY` method definition.

When you deploy this template were you able to trigger the function?

Thanks

**Comment by txase** [comment]
@mndeveci Yes, with the permissions set with the second * in the arn, invocations from the http api fail. When the extra * path is removed, it starts working.

**Comment by hsiaoa** [comment]
+1
Reporting the same exact issue.

**Comment by mndeveci** [comment]
Thanks @txase & @hsiaoa ,

So I think this is related to `/$default` path. If you define another path (ex: `/hello`) then current behaviour is working as expected;

So SAM Translator should transform following template;
```
Transform: AWS::Serverless-2016-10-31
Resources:
  Function:
    Type: AWS::Serverless::Function
    Properties:
      InlineCode: 'module.exports.handler = () => ({})'
      Handler: index.handler
      Runtime: nodejs12.x
      Events:
        HttpApiANYdefault:
          Type: HttpApi
          Properties:
            Path: /$default
            Method: ANY
            ApiId: !Ref HttpApi
  Function2:
    Type: AWS::Serverless::Function
    Properties:
      InlineCode: 'module.exports.handler = () => ({})'
      Handler: index.handler
      Runtime: nodejs12.x
      Events:
        HttpApiANYdefault:
          Type: HttpApi
          Properties:
            Path: /hello
            Method: ANY
            ApiId: !Ref HttpApi
  HttpApi:
    Type: AWS::Serverless::HttpApi
    Properties:
      DefinitionBody:
        openapi: '3.0'
        info:
          title: !Sub ${AWS::StackName}-HttpApi
          version: '1.0'
        paths:
          /$default:
            x-amazon-apigateway-any-method:
              responses: {}
              isDefaultRoute: true
          /hello:
            x-amazon-apigateway-any-method:
              responses: { }
      FailOnWarnings: true
```
Into this one;

```
{
  "Resources": {
    "Function": {
      "Type": "AWS::Lambda::Function", 
      "Properties": {
        "Handler": "index.handler", 
        "Code": {
          "ZipFile": "module.exports.handler = () => ({})"
        }, 
        "Role": {
          "Fn::GetAtt": [
            "FunctionRole", 
            "Arn"
          ]
        }, 
        "Runtime": "nodejs12.x", 
        "Tags": [
          {
            "Value": "SAM", 
            "Key": "lambda:createdBy"
          }
        ]
      }
    }, 
    "FunctionHttpApiANYdefaultPermission": {
      "Type": "AWS::Lambda::Permission", 
      "Properties": {
        "Action": "lambda:InvokeFunction", 
        "Principal": "apigateway.amazonaws.com", 
        "FunctionName": {
          "Ref": "Function"
        }, 
        "SourceArn": {
          "Fn::Sub": [
            "arn:${AWS::Partition}:execute-api:${AWS::Region}:${AWS::AccountId}:${__ApiId__}/${__Stage__}/$default", 
            {
              "__Stage__": "*", 
              "__ApiId__": {
                "Ref": "HttpApi"
              }
            }
          ]
        }
      }
    }, 
    "FunctionRole": {
      "Type": "AWS::IAM::Role", 
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Version": "2012-10-17", 
          "Statement": [
            {
              "Action": [
                "sts:AssumeRole"
              ], 
              "Effect": "Allow", 
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com"
                ]
              }
            }
          ]
        }, 
        "ManagedPolicyArns": [
          "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
        ], 
        "Tags": [
          {
            "Value": "SAM", 
            "Key": "lambda:createdBy"
          }
        ]
      }
    }, 
    "Function2HttpApiANYdefaultPermission": {
      "Type": "AWS::Lambda::Permission", 
      "Properties": {
        "Action": "lambda:InvokeFunction", 
        "Principal": "apigateway.amazonaws.com", 
        "FunctionName": {
          "Ref": "Function2"
        }, 
        "SourceArn": {
          "Fn::Sub": [
            "arn:${AWS::Partition}:execute-api:${AWS::Region}:${AWS::AccountId}:${__ApiId__}/${__Stage__}/*/hello", 
            {
              "__Stage__": "*", 
              "__ApiId__": {
                "Ref": "HttpApi"
              }
            }
          ]
        }
      }
    }, 
    "Function2": {
      "Type": "AWS::Lambda::Function", 
      "Properties": {
        "Handler": "index.handler", 
        "Code": {
          "ZipFile": "module.exports.handler = () => ({})"
        }, 
        "Role": {
          "Fn::GetAtt": [
            "Function2Role", 
            "Arn"
          ]
        }, 
        "Runtime": "nodejs12.x", 
        "Tags": [
          {
            "Value": "SAM", 
            "Key": "lambda:createdBy"
          }
        ]
      }
    }, 
    "HttpApi": {
      "Type": "AWS::ApiGatewayV2::Api", 
      "Properties": {
        "Body": {
          "info": {
            "version": "1.0", 
            "title": {
              "Fn::Sub": "${AWS::StackName}-HttpApi"
            }
          }, 
          "paths": {
            "/$default": {
              "x-amazon-apigateway-any-method": {
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST", 
                  "type": "aws_proxy", 
                  "uri": {
                    "Fn::Sub": "arn:${AWS::Partition}:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${Function.Arn}/invocations"
                  }, 
                  "payloadFormatVersion": "2.0"
                }, 
                "isDefaultRoute": true, 
                "responses": {}
              }
            }, 
            "/hello": {
              "x-amazon-apigateway-any-method": {
                "x-amazon-apigateway-integration": {
                  "httpMethod": "POST", 
                  "type": "aws_proxy", 
                  "uri": {
                    "Fn::Sub": "arn:${AWS::Partition}:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${Function2.Arn}/invocations"
                  }, 
                  "payloadFormatVersion": "2.0"
                }, 
                "responses": {}
              }
            }
          }, 
          "openapi": "3.0", 
          "tags": [
            {
              "name": "httpapi:createdBy", 
              "x-amazon-apigateway-tag-value": "SAM"
            }
          ]
        }, 
        "FailOnWarnings": true
      }
    }, 
    "HttpApiApiGatewayDefaultStage": {
      "Type": "AWS::ApiGatewayV2::Stage", 
      "Properties": {
        "ApiId": {
          "Ref": "HttpApi"
        }, 
        "AutoDeploy": true, 
        "StageName": "$default", 
        "Tags": {
          "httpapi:createdBy": "SAM"
        }
      }
    }, 
    "Function2Role": {
      "Type": "AWS::IAM::Role", 
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Version": "2012-10-17", 
          "Statement": [
            {
              "Action": [
                "sts:AssumeRole"
              ], 
              "Effect": "Allow", 
              "Principal": {
                "Service": [
                  "lambda.amazonaws.com"
                ]
              }
            }
          ]
        }, 
        "ManagedPolicyArns": [
          "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
        ], 
        "Tags": [
          {
            "Value": "SAM", 
            "Key": "lambda:createdBy"
          }
        ]
      }
    }
  }
}
```

**Comment by aaythapa** [comment]
Closing this as it was resolved in #2574
